internal/expr/prepare.go

```
@@ -0,0 +1,183 @@
+func prepareInput(ti typeNameToInfo, p *inputPart) error {
+	if inf, ok := ti[p.source.prefix]; ok {
+		if _, ok := inf.tagToField[p.source.name]; ok {
+			return nil
+		}
+		return fmt.Errorf(`there is no tag with name "%s" in "%s"`,
+			p.source.name, inf.structType.Name())
+	}
+	return fmt.Errorf(`unknown type: "%s"`, p.source.prefix)
+}
+
+func starCount(fns []fullName) int {
+	s := 0
+	for _, fn := range fns {
+		if fn.name == "*" {
+			s++
+		}
+	}
+	return s
+}
+
+func prepareOutput(ti typeNameToInfo, p *outputPart) ([]string, error) {
+
+	var outCols []string = make([]string, 0)
+
+	// Check target struct type and its tags are valid.
+	for _, t := range p.target {
+		inf, ok := ti[t.prefix]
+		if !ok {
+			return nil, fmt.Errorf("unknown type: %s", t.prefix)
+		}
+
+		_, ok = inf.tagToField[t.name]
+		if !ok && t.name != "*" {
+			return nil, fmt.Errorf(`there is no tag with name "%s" in "%s"`, t.name, inf.structType.Name())
+		}
+	}
+
+	// Check asterisk are in correct places.
+
+	sct := starCount(p.target)
+	scc := starCount(p.source)
+
+	if sct > 1 || scc > 1 || (scc == 1 && sct == 0) {
+		return nil, fmt.Errorf("invalid asterisk in output expression")
+	}
+
+	starTarget := sct == 1
+	starColumn := scc == 1
+
+	lenS := len(p.source)
+	lenT := len(p.target)
+
+	if (starTarget && lenT > 1) || (starColumn && lenS > 1) {
+		return nil, fmt.Errorf("invalid mix of asterisk and none asterisk columns in output expression")
+	}
+
+	if !starTarget && (lenS > 0 && (lenT != lenS)) {
+		return nil, fmt.Errorf("mismatched number of cols and targets in output expression")
+	}
+
+	// Case 1: Star target cases e.g. "...&P.*".
+	// In parse we ensure that if p.target[0] is a * then len(p.target) == 1
+	if starTarget {
+
+		inf, _ := ti[p.target[0].prefix]
+
+		// Case 1.1: Single star e.g. "t.* AS &P.*" or "&P.*"
+		if starColumn || lenS == 0 {
+			pref := ""
+
+			// Prepend table name. E.g. "t" in "t.* AS &P.*".
+			if lenS > 0 && p.source[0].prefix != "" {
```

Do we need to return any error if `lenS >0` but there's no prefix? Or has it been checked at this point?

---

Ah yes, I think this should actually be `if starSource`